src/cross_section.F90

```diff
@@ -55,32 +54,23 @@ subroutine calculate_xs(p)
       ! Determine if this material has S(a,b) tables
       check_sab = (mat % n_sab > 0)
 
-      ! Initialize position in i_sab_nuclides
-      j = 1
-
       ! Add contribution from each nuclide in material
       do i = 1, mat % n_nuclides
         ! ========================================================================
         ! CHECK FOR S(A,B) TABLE
 
         i_sab = 0
 
-        ! Check if this nuclide matches one of the S(a,b) tables specified -- this
-        ! relies on i_sab_nuclides being in sorted order
+        ! Check if this nuclide matches one of the S(a,b) tables specified
         if (check_sab) then
-          if (i == mat % i_sab_nuclides(j)) then
+          if (mat % nuc_sab_dict % has_key(i)) then
```

Comment:
The dictionary indeed has potential performance issue, considering it has so small size. I'm not sure how bad will it be. I'm trying to make some comparisons.
In addition, an alternative to the dictionary approach is to use a nuc-to-sab array which has the same size of nuclide array in a material. The sab index can be fetched directly using the nuclide index, like this:

``` forrtan
do i = 1, mat % n_nuclides
  i_sab = mat % sab_index_of_nuc(i)
```
This approach should have better performance than the current comparison method but may need a bit more memory.